GeoJSON Upload Fixes

rajadain opened the pull request:
## Overview

Fixes a number of issues encountered during GeoJSON upload. See commit messages for details.

Connects #2481 
Connects #2483 

### Demo

![2017-11-08 15 21 49](https://user-images.githubusercontent.com/1430060/32572722-7ec58206-c499-11e7-91d9-9bd872961cf3.gif)

### Notes

We should do more validation of the GeoJSON file (for example, ensure it contains a MultiPolygon or Polygon, and not just a Point or Line, which would be valid GeoJSON but invalid for our purposes), and there is a probably a library we can use for that. I'm leaving that for a later date. See #2499.

## Testing Instructions

Download these test files, and rename them to end in `.geojson`:

 * [feature.geojson.txt](https://github.com/WikiWatershed/model-my-watershed/files/1455321/feature.geojson.txt)
 * [featurecollection.geojson.txt](https://github.com/WikiWatershed/model-my-watershed/files/1455322/featurecollection.geojson.txt)
 * [multipolygon.geojson.txt](https://github.com/WikiWatershed/model-my-watershed/files/1455324/multipolygon.geojson.txt)
 * [polygon.geojson.txt](https://github.com/WikiWatershed/model-my-watershed/files/1455325/polygon.geojson.txt)
 * [parse-fail.geojson.txt](https://github.com/WikiWatershed/model-my-watershed/files/1455326/parse-fail.geojson.txt)

Ensure that all of them upload and do not crash the app.
Ensure that the first four upload correctly and show the shape.
Ensure that the last one shows an error message.

Try any other shapes you can.

kellyi commented:
Taking a look at this.

kellyi commented:
Linter error here is causing the build to be unstable:

```
js/src/draw/utils.js: line 97, col 21, 'nesting' is already defined.
```

kellyi commented:
Double checking: should this be targeted at `develop` and not `release`?

ajrobbins commented:
yep, develop!

kellyi commented on src/mmw/js/src/draw/views.js:
```diff
@@ -411,9 +422,25 @@ var AoIUploadView = Marionette.ItemView.extend({
     handleGeoJSON: function(jsonString) {
-        var geojson = JSON.parse(jsonString);
+        var geojson = JSON.parse(jsonString),
+            polygon = _.has(geojson, 'coordinates') ? geojson :
```
Consider replacing these chained ternaries with an if block or an IIFE like:

```
polygon = (function() {
    if (_.has(geojson, 'coordinates') {
        return geojson;
    } else if (_.has(geojson, 'geometry') && geojson.geometry) {
        return geojson.geometry;
    } else if (.....
    ....
    }
    return null;
})();
```

kellyi commented:
Tested the json files posted above and each worked as expected. Going to try a few other shapes now, too.

kellyi commented:
Generated this RWD shape through Swagger and it worked:

[map (1).geojson.txt](https://github.com/WikiWatershed/model-my-watershed/files/1458040/map.1.geojson.txt)

![screen shot 2017-11-09 at 10 17 32 am](https://user-images.githubusercontent.com/4165523/32612817-3a5ca8da-c537-11e7-8ef7-8c5c71d90ad3.png)

kellyi commented:
Tried testing out this one https://raw.githubusercontent.com/johan/world.geo.json/master/countries/USA/PA/Allegheny.geo.json and it reported an error. Checked it with http://geojsonlint.com/ and the file's invalid, so the error's correct.

rajadain commented:
Thanks for the suggestions. I moved the complex ternary clause into a utility function, fixed the lint, and added some GeoJSON parsing tests. Tested locally to ensure it still works. Will squash and merge once CI is green.

rajadain commented:
Thanks for taking a look!
